### PR TITLE
feat(ec): ability to upgrade the cluster

### PR DESCRIPTION
**Pull request by emosbaugh**
#### What this PR does / why we need it:

Original PR https://github.com/replicatedhq/kots/pull/4678

Related https://github.com/replicatedhq/embedded-cluster-operator/pull/218

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/embeddedcluster/oras_test.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "test1",
+			args: args{
+				ctx:     context.Background(),
+				srcRepo: "ttl.sh/ethan/embedded-cluster-operator-bin:24h",
```
Comment (sgalsaleh) [inline pkg/embeddedcluster/oras_test.go:23]: is this what we want to use for testing?

File: pkg/embeddedcluster/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+		}
+		args = append(args, "--local-artifact-mirror-image", localArtifactMirrorImage)
+	}
+	args = append(args, "--installation", "-")
+
+	cmd := exec.CommandContext(ctx, bin, args...)
+	cmd.Stdin = strings.NewReader(string(installationData))
+	pr, pw := io.Pipe()
+	defer pw.Close()
+	cmd.Stdout = pw
```
Comment (sgalsaleh) [inline pkg/embeddedcluster/upgrade.go:158]: why not just 
```suggestion
	cmd.Stdout = os.Stdout
```

File: pkg/embeddedcluster/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+// operator, wait for the CRD to be up-to-date, and then apply the installation object.
+func runClusterUpgrade(
+	ctx context.Context, k8sClient kubernetes.Interface,
+	in *embeddedclusterv1beta1.Installation, registrySettings registrytypes.RegistrySettings,
+	license kotsv1beta1.License, versionLabel string,
+) error {
+	var bin string
+
+	if in.Spec.AirGap {
+		image := in.Spec.Artifacts.AdditionalArtifacts["operator"]
```
Comment (sgalsaleh) [inline pkg/embeddedcluster/upgrade.go:109]: `image` is a bit confusing. how about `location`?
```suggestion
		location := in.Spec.Artifacts.AdditionalArtifacts["operator"]
```

File: pkg/embeddedcluster/upgrade.go [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+		if image == "" {
+			return fmt.Errorf("missing operator image in airgap artifacts")
```
Comment (sgalsaleh) [inline pkg/embeddedcluster/upgrade.go:111]:
```suggestion
			return fmt.Errorf("missing operator binary in airgap artifacts")
```

File: pkg/embeddedcluster/oras_test.go (hunk 1 above)
Comment (emosbaugh) [reply pkg/embeddedcluster/oras_test.go:23]: the test is prefixed with an x_ so it does not run. do you want me to remove it?

File: pkg/embeddedcluster/upgrade.go (hunk 2 above)
Comment (emosbaugh) [reply pkg/embeddedcluster/upgrade.go:158]: this is equivalent. why would i change it?

File: pkg/embeddedcluster/oras_test.go (hunk 1 above)
Comment (sgalsaleh) [reply pkg/embeddedcluster/oras_test.go:23]: if it does not run, then we should probably just remove it yeah.

File: pkg/embeddedcluster/upgrade.go (hunk 2 above)
Comment (sgalsaleh) [reply pkg/embeddedcluster/upgrade.go:158]: i meant for simplicity without having the below goroutine entirely.

Comment (sgalsaleh) [reply pkg/embeddedcluster/upgrade.go:158]: something like this https://github.com/replicatedhq/kots/blob/kots-upgrader/pkg/upgradeservice/process.go#L95-L96

Comment (sgalsaleh) [reply pkg/embeddedcluster/upgrade.go:158]: I guess your way has the benefit of distinguishing it's the binary's output

**Review by sgalsaleh — APPROVED** [review]

File: pkg/embeddedcluster/upgrade.go [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+	kotsv1beta1 "github.com/replicatedhq/kotskinds/apis/kots/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/credentials"
+	k8syaml "sigs.k8s.io/yaml"
+)
+
+const (
+	clusterUpgradeTimeout = 5 * time.Minute
```
Comment (sgalsaleh) [inline pkg/embeddedcluster/upgrade.go:35]: Is this timeout sufficient to upgrade the operator chart and run the LAM jobs specially on a large number of nodes?

Comment (emosbaugh) [reply pkg/embeddedcluster/upgrade.go]: ive increased it to 15 minutes

File: pkg/embeddedcluster/upgrade.go [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+
+// startClusterUpgrade will create a new installation with the provided config.
+func startClusterUpgrade(
+	ctx context.Context, newcfg embeddedclusterv1beta1.ConfigSpec,
+	artifacts *embeddedclusterv1beta1.ArtifactsLocation,
+	registrySettings registrytypes.RegistrySettings,
+	license kotsv1beta1.License, versionLabel string,
+) error {
+	// TODO(upgrade): put a lock here to prevent multiple upgrades at the same time
+
```
Comment (laverya) [inline pkg/embeddedcluster/upgrade.go:46]: Global mutex? Should be easy, no?

Comment (laverya) [reply pkg/embeddedcluster/upgrade.go:46]: DB lock would be better but then you have issues with failures

File: pkg/embeddedcluster/upgrade.go [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+			ClusterID:                 current.Spec.ClusterID,
+			MetricsBaseURL:            current.Spec.MetricsBaseURL,
+			HighAvailability:          current.Spec.HighAvailability,
+			AirGap:                    current.Spec.AirGap,
+			Artifacts:                 artifacts,
+			Config:                    &newcfg,
+			EndUserK0sConfigOverrides: current.Spec.EndUserK0sConfigOverrides,
+			BinaryName:                current.Spec.BinaryName,
+			LicenseInfo:               &embeddedclusterv1beta1.LicenseInfo{IsDisasterRecoverySupported: license.Spec.IsDisasterRecoverySupported},
+		},
```
Comment (laverya) [inline pkg/embeddedcluster/upgrade.go:82]: At some point we should start doing spec = current.spec and then just replacing the artifacts and configs

**Review by laverya — APPROVED** [review]

**Review by sgalsaleh — APPROVED** [review]

**Pull request merged**